diff --git a/include/Marker.h b/include/Marker.h
index cf34750..2745b88 100644
--- a/include/Marker.h
+++ b/include/Marker.h
@@ -20,6 +20,7 @@
 #include <vector>
 #include <map>
 #include <utility>
+#include <cstdint>
 using std::vector;
 using std::string;
 using std::to_string;
diff --git a/include/Pheno.h b/include/Pheno.h
index 9241d07..2d43fa9 100644
--- a/include/Pheno.h
+++ b/include/Pheno.h
@@ -19,6 +19,7 @@
 #include <vector>
 #include <string>
 #include <map>
+#include <cstdint>
 using std::map;
 using std::vector;
 using std::string;
diff --git a/include/utils.hpp b/include/utils.hpp
index 70045e0..e0705c5 100644
--- a/include/utils.hpp
+++ b/include/utils.hpp
@@ -31,6 +31,7 @@
 #include <numeric>
 #include <algorithm>
 #include <sstream>
+#include <cstdint>
 
 std::string getHostName();
 std::string getLocalTime();
