From a686e5d40fc5f6b54be386589d17e8ff57f8f8ee Mon Sep 17 00:00:00 2001
From: Lea Waller <lea.waller@charite.de>
Date: Thu, 12 May 2022 22:16:26 +0200
Subject: Install packages to default R library directory

---
 CMakeLists.txt | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index bddbb9a..5c043d8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -33,7 +33,7 @@ add_custom_target(epactsR-install
                   ALL
                   DEPENDS epactsR-build
                   #${CMAKE_COMMAND} -E env CPLUS_INCLUDE_PATH=${CMAKE_SOURCE_DIR}/cget/include LIBRARY_PATH=${CMAKE_SOURCE_DIR}/cget/lib R CMD INSTALL -l ${CMAKE_BINARY_DIR} epactsR_3.3.3.tar.gz
-                  COMMAND R CMD INSTALL -l ${CMAKE_BINARY_DIR} --configure-vars='INCLUDE_DIR=${CGET_PREFIX}/include LINK_DIR=${CGET_PREFIX}/lib' epactsR_${PROJECT_VERSION}.tar.gz
+                  COMMAND R CMD INSTALL epactsR_${PROJECT_VERSION}.tar.gz
                   WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
                   COMMENT "Installing epactsR ...")
 
@@ -44,7 +44,7 @@ add_custom_target(mmSKAT-build
 add_custom_target(mmSKAT-install
                   ALL
                   DEPENDS mmSKAT-build
-                  COMMAND R CMD INSTALL -l ${CMAKE_BINARY_DIR} mmSKAT_0.95.tar.gz
+                  COMMAND R CMD INSTALL mmSKAT_0.95.tar.gz
                   WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
                   COMMENT "Installing mmSKAT ...")
 
@@ -88,12 +88,10 @@ install(FILES ${EPACTS_PERL_SCRIPTS} ${CMAKE_BINARY_DIR}/epacts-version
         DESTINATION bin)
 install(FILES ${EPACTS_DATA_FILES} DESTINATION share/EPACTS)
 install(TARGETS pEmmax vcfast chaps anno bgzip tabix parse-sample-ids RUNTIME DESTINATION bin)
-install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib)
-install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib)
 
 set(CPACK_PACKAGE_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
 set(CPACK_PACKAGE_VERSION_MINOR ${PROJECT_VERSION_MINOR})
 set(CPACK_PACKAGE_VERSION_PATCH ${PROJECT_VERSION_PATCH})
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "ghostscript, r-base, groff, gnuplot") 
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "ghostscript, r-base, groff, gnuplot")
 set(CPACK_RPM_PACKAGE_REQUIRES "ghostscript, R, groff, gnuplot")
 include(CPack)
-- 
2.36.1

