diff --git a/setup.py b/setup.py
index 0ca60d2..ed30993 100644
--- a/setup.py
+++ b/setup.py
@@ -6,6 +6,7 @@
 # LICENSE.txt file in the root directory of this source tree)
 #######################################################################

+import os
 from pathlib import Path
 from textwrap import dedent

@@ -62,5 +63,7 @@ setup(
     platforms=["any"],
     packages=["blosc2"],
     package_dir={"blosc2": "blosc2"},
+    cmake_args=["-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON"]
+    + (["-DUSE_SYSTEM_BLOSC2:BOOL=ON"] if int(os.environ.get("USE_SYSTEM_BLOSC2", "0")) else []),
     install_requires=open("requirements-runtime.txt").read().split(),
 )
