From 08e48b496c22ad10bdb995bd9e3550bddfc774fc Mon Sep 17 00:00:00 2001
From: Lea Waller <lea.waller@charite.de>
Date: Thu, 12 May 2022 22:47:25 +0200
Subject: Do not install `tabix`

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index bddbb9a..fb4e261 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -87,13 +87,13 @@ install(FILES ${EPACTS_PERL_SCRIPTS} ${CMAKE_BINARY_DIR}/epacts-version
         PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
         DESTINATION bin)
 install(FILES ${EPACTS_DATA_FILES} DESTINATION share/EPACTS)
-install(TARGETS pEmmax vcfast chaps anno bgzip tabix parse-sample-ids RUNTIME DESTINATION bin)
+install(TARGETS pEmmax vcfast chaps anno parse-sample-ids RUNTIME DESTINATION bin)
 install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib)
 install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib)
 
 set(CPACK_PACKAGE_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
 set(CPACK_PACKAGE_VERSION_MINOR ${PROJECT_VERSION_MINOR})
 set(CPACK_PACKAGE_VERSION_PATCH ${PROJECT_VERSION_PATCH})
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "ghostscript, r-base, groff, gnuplot") 
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "ghostscript, r-base, groff, gnuplot")
 set(CPACK_RPM_PACKAGE_REQUIRES "ghostscript, R, groff, gnuplot")
 include(CPack)
-- 
2.36.1

