From 49d31fb08fef47c9bfa6a61387ecfa97cff7fc24 Mon Sep 17 00:00:00 2001
From: Lea Waller <lea.waller@charite.de>
Date: Thu, 12 May 2022 22:16:26 +0200
Subject: Install packages to default R library directory

---
 CMakeLists.txt | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index fb4e261..3ea3b11 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -33,7 +33,7 @@ add_custom_target(epactsR-install
                   ALL
                   DEPENDS epactsR-build
                   #${CMAKE_COMMAND} -E env CPLUS_INCLUDE_PATH=${CMAKE_SOURCE_DIR}/cget/include LIBRARY_PATH=${CMAKE_SOURCE_DIR}/cget/lib R CMD INSTALL -l ${CMAKE_BINARY_DIR} epactsR_3.3.3.tar.gz
-                  COMMAND R CMD INSTALL -l ${CMAKE_BINARY_DIR} --configure-vars='INCLUDE_DIR=${CGET_PREFIX}/include LINK_DIR=${CGET_PREFIX}/lib' epactsR_${PROJECT_VERSION}.tar.gz
+                  COMMAND R CMD INSTALL epactsR_${PROJECT_VERSION}.tar.gz
                   WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
                   COMMENT "Installing epactsR ...")
 
@@ -44,7 +44,7 @@ add_custom_target(mmSKAT-build
 add_custom_target(mmSKAT-install
                   ALL
                   DEPENDS mmSKAT-build
-                  COMMAND R CMD INSTALL -l ${CMAKE_BINARY_DIR} mmSKAT_0.95.tar.gz
+                  COMMAND R CMD INSTALL mmSKAT_0.95.tar.gz
                   WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
                   COMMENT "Installing mmSKAT ...")
 
@@ -88,8 +88,6 @@ install(FILES ${EPACTS_PERL_SCRIPTS} ${CMAKE_BINARY_DIR}/epacts-version
         DESTINATION bin)
 install(FILES ${EPACTS_DATA_FILES} DESTINATION share/EPACTS)
 install(TARGETS pEmmax vcfast chaps anno parse-sample-ids RUNTIME DESTINATION bin)
-install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/epactsR DESTINATION lib)
-install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib) #install(DIRECTORY ${CMAKE_BINARY_DIR}/mmSKAT DESTINATION lib)
 
 set(CPACK_PACKAGE_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
 set(CPACK_PACKAGE_VERSION_MINOR ${PROJECT_VERSION_MINOR})
-- 
2.36.1


From 6474cf8b181c16250250b1692f603637d34a4048 Mon Sep 17 00:00:00 2001
From: Lea Waller <lea.waller@charite.de>
Date: Thu, 12 May 2022 23:51:19 +0200
Subject: Remove custom `require` paths

---
 data/epactsMulti.R   | 6 ++----
 data/epactsSingle.R  | 6 ++----
 scripts/epacts-group | 3 +--
 3 files changed, 5 insertions(+), 10 deletions(-)

diff --git a/data/epactsMulti.R b/data/epactsMulti.R
index 8106585..3a19604 100644
--- a/data/epactsMulti.R
+++ b/data/epactsMulti.R
@@ -20,11 +20,9 @@ test <- args[17]
 setwd(paste(bindir,"/share/EPACTS",sep=""))
 
 ### Wrapper R script to perform single variant analysis
-if ( !require(epactsR,lib.loc=paste(bindir,"/lib/",sep="") ) ) {
-  stop(paste("Cannot find epactsR package at ",bindir,"/lib/",sep=""))
+if ( !require(epactsR ) ) {
+  stop(paste("Cannot find epactsR package"))
 }
-## the DLL file has to be loaded explictly for some machine - don't know why
-dyn.load(paste(bindir,"/lib/epactsR/libs/epactsR",.Platform$dynlib.ext, sep=""))
 
 pnames <- scan(phenof,what=character(),nlines=1)[-1]
 phenos <- as.matrix(read.table(phenof)[,-1])
diff --git a/data/epactsSingle.R b/data/epactsSingle.R
index 9d9c60a..4f53392 100644
--- a/data/epactsSingle.R
+++ b/data/epactsSingle.R
@@ -39,11 +39,9 @@ if ( test == "single.q.emmaxR" ) {  ## special treatment for emmaxR
 setwd(paste(bindir,"/share/EPACTS",sep=""))
 
 ### Wrapper R script to perform single variant analysis
-if ( !require(epactsR,lib.loc=paste(bindir,"/lib/",sep="") ) ) {
-  stop(paste("Cannot find epactsR package at ",bindir,"/lib/",sep=""))
+if ( !require(epactsR) ) {
+  stop(paste("Cannot find epactsR package"))
 }
-## the DLL file has to be loaded explictly for some machine - don't know why
-dyn.load(paste(bindir,"/lib/epactsR/libs/epactsR",.Platform$dynlib.ext, sep=""))
   
 pheno <- as.double(read.table(phenof)[,2])
 pheno <- pheno - min(pheno,na.rm=T)
diff --git a/scripts/epacts-group b/scripts/epacts-group
index 57577c6..16e678e 100755
--- a/scripts/epacts-group
+++ b/scripts/epacts-group
@@ -412,10 +412,9 @@ else {
 	    print R "binaryFlag <- ".($binaryFlag ? "TRUE" : "FALSE")."\n";
 	    print R "setwd('$datadir')\n";
 	    print R "batch <- $i\n";
-	    print R "if ( !require(epactsR,lib.loc=paste(\"$epactsdir\",\"/lib/\",sep=\"\") ) ) {\n";
+	    print R "if ( !require(epactsR) ) {\n";
 	    print R "\tstop(\"Cannot find epactsR package\")\n";
 	    print R "}\n";
-	    print R "dyn.load(paste(\"$epactsdir/lib/epactsR/libs/epactsR\",.Platform\$dynlib.ext, sep=\"\"))\n";
 	    print R "bindir <- '$epactsdir'\n";
 	    print R "source('epactsGroup.R')\n";
 	    close R;
-- 
2.36.1

